### ISSUE TITLE
Add verbose diagnostic for opening xrootd file

### ISSUE THREAD

**Issue by vkuznet** [scikit-hep/uproot#148]
Jim,
I have two issues with opening XrootD files and it would be helpful if you'll add some verbose output about steps involved in access to XrootD server.

Problem 1: I'm using GPU node (at Cincinatti) and over there when I do uproot.open call it just sits there doing nothing. I don't know what's wrong and I can't figure out where to look at. The file I'm trying to open is
root://cms-xrd-global.cern.ch//store/data/Run2018C/Tau/NANOAOD/14Sep2018_ver3-v1/60000/6FA4CC7C-8982-DE4C-BEED-C90413312B35.root

Problem 2: on my mac I can't open this file either, but I get some output which I don't know how to interpret:
```
>>> data=uproot.open(fin)
Traceback (most recent call last):
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/cachetools/cache.py", line 51, in __getitem__
    return self.__data[key]
KeyError: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/uproot/source/chunked.py", line 84, in data
    chunk = self.cache[chunkindex]
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/uproot/cache.py", line 79, in __getitem__
    return self._cache[where]
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/cachetools/lru.py", line 16, in __getitem__
    value = cache_getitem(self, key)
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/cachetools/cache.py", line 53, in __getitem__
    return self.__missing__(key)
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/cachetools/cache.py", line 80, in __missing__
    value = self.__missing(key)
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/cachetools/cache.py", line 110, in __missing
    raise KeyError(key)
KeyError: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/uproot/rootio.py", line 75, in open
    return xrootd(path, xrootdsource)
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/uproot/rootio.py", line 88, in xrootd
    return ROOTDirectory.read(openfcn(path), **options)
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/uproot/rootio.py", line 127, in read
    magic, fVersion = cursor.fields(source, ROOTDirectory._format1)
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/uproot/source/cursor.py", line 70, in fields
    return format.unpack(source.data(start, stop))
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/uproot/source/chunked.py", line 87, in data
    chunk = self.cache[chunkindex] = self._read(chunkindex)
  File "/opt/anaconda3/envs/fastai-cpu/lib/python3.6/site-packages/uproot/source/xrootd.py", line 80, in _read
    status, data = self._source.read(chunkindex * self._chunkbytes, self._chunkbytes)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xef in position 6: invalid continuation byte
```

Meanwhile I can open this file on lxplus without any issues. My main concern is to figure out what's wrong with Problem 1 where I uproot hang and waits for something.

**Comment by jpivarski** [comment]
Problem 1, hanging on open: this must be in one of the `pyxrootd.client.File` calls— `open`, `status`, `read`, or `close` (those are the only ones I make). Immediately after each call, I do

```python
if status.get("error", None):
    raise OSError(status["message"])
```

so if the pyxrootd call itself doesn't hang, we get to the exception right away. I could look into setting timeouts for the pyxrootd calls, which would at least provide an error message after N seconds.

By the way, I opened this file in uproot without issues.

**Comment by jpivarski** [comment]
Problem 2, the reason for the `KeyError: 0` is because I keep a cache of chunks of the file. It's a chained exception; the only meaningful one is the last one about the UnicodeDecodeError.

This is again in pyxrootd call, `pyxrootd.client.File.read`. I'm very surprised that there's any unicode business going on there— this is supposed to be a binary file. I wonder if `read` is defined in C, and therefore the call stack ends here though the error is being raised by C-API calls within pyxrootd. I'll investigate.

That said, there might be an issue with your XRootD installation. Mine (which works for this file) is 4.8.4.

**Comment by jpivarski** [comment]
The `read` method is, indeed, [defined in a C extension](https://github.com/xrootd/xrootd/blob/cfdc6aca0bde74f7b65000987f846512c42a87ab/bindings/python/src/PyXRootDFile.cc#L140) so that's why the call stack ends on this line. I don't see anything in here about unicode, but maybe that's in the [Python 2 → 3 conversion](https://github.com/xrootd/xrootd/commit/d74d5142da7c700f149e90662bdccb53fbc5502f#diff-3eef6cce7899534a6d247bb0ed4f8df0)?

These methods do have a keyword argument `timeout`, so I could take such an option in `uproot.open` and propagate it down to each pyxrootd call. That, at least, would turn hanging connections into exceptions.

**Comment by vkuznet** [comment]
thanks, that would be helpful.

**Comment by chrisburr** [comment]
1) You control the output level of XRootD by running `export XRD_LOGLEVEL=Debug` before starting Python. You can also reduce the timeout duration and number of retries used by setting environment variables as documented [here](http://xrootd.org/doc/man/xrdcp.1.html#ENVIRONMENT).

2) This looks like you're using v4.7.x of XRootD. If so, there is a bug in the Python 3 bindings which makes reading files impossible which is fixed in v4.8+ ([see here](https://github.com/xrootd/xrootd/issues/632)).

**Comment by matthewfeickert** [comment]
> ...in part to test the new deployment script 

I actually removed the parts that changed deployment in PR #149 to keep the scope of the PR focused on Binder. I'll add those in a new PR tomorrow though.

**Comment by vkuznet** [comment]
Chris,
thanks for hints. At least now I see output of xrootd flow and posted it on
xrootd github. And, I solved my problem 2 by upgrading to newer version.
Valentin.

**Comment by jpivarski** [comment]
And this is an example of use (written on #155, but should be repeated here):

```python
import uproot
f = uproot.open("root://wherever//whatever.root", xrootdsource={"timeout": 1})
```
